File: LowLevelDesign/tic-tac-toe/src/Scoreboard.ts
import { Game } from "./Game";
import { GameObserver } from "./observer/GameObserver";

export class Scoreboard implements GameObserver {
    private scores: Map<string, number>;

    constructor() {
        this.scores = new Map<string, number>();
    }

    public update(game: Game): void {
        const winner = game.getWinner();
        if (winner) {
            const currentScore = this.scores.get(winner.name) || 0;
            this.scores.set(winner.name, currentScore + 1);
            console.log(`Scoreboard Updated: ${winner.name} has ${this.scores.get(winner.name)} wins.`);
        } else {
            console.log("No winner yet, scoreboard remains unchanged.");
        }
    }

}

File: LowLevelDesign/java/tictactoe/src/com/tictac/observer/Scoreboard.java
package com.tictac.observer;

import com.tictac.entities.Game;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public class Scoreboard implements GameObserver{
    private Map<String, Integer> scores;

    public Scoreboard() {
        this.scores =  new ConcurrentHashMap<>();
    }

    public void update(Game game) {
        if(game.getWinner() != null) {
            String winnerName = game.getWinner().getName();
            this.scores.put(winnerName, scores.getOrDefault(winnerName, 0) + 1);
            System.out.printf("[Scoreboard] %s wins! Their new score is %d.%n", winnerName, scores.get(winnerName));
        }
    }
}
